src/mo_frontend/bi_match.ml

```diff
@@ -153,28 +159,23 @@ end = struct
   type t = (typ * con * typ) list ref
   let empty () = ref []
   let add t lb c ub = t := (lb, c, ub) :: !t
-  let to_string t = 
+
+  let to_string t =
     let parts = List.rev !t in
     if parts = [] then "" else
+    let s = if List.length parts > 1 then "s" else "" in
     Format.asprintf
-      "cannot solve invariant type parameter%s %s, no principal solution with%a\nwhere%a"
-      (if List.length parts > 1 then "s" else "")
-      (String.concat ", " (List.map (fun (_, c, _) -> Cons.name c) parts))
-      display_constraints parts
-      display_rels (List.map (fun (lb, _, ub) -> (lb,"=/=",ub)) parts)
+      "there is no \"best\" choice for type parameter%s `%s`."
+      s
+      (String.concat "`, `" (List.map (fun (_, c, _) -> Cons.name c) parts))
+
 end
 
+
 let fail_over_constrained lb c ub =
-(*  if debug then *)
-    error (Format.asprintf
-      "implicit instantiation of type parameter `%s` is over-constrained with%a\nwhere%a\nso that no valid instantiation exists"
-      (Cons.name c)
-      display_constraint (lb, c, ub)
-      display_rel (lb, "</:", ub))
-(*  else
-    raise (Bimatch (Format.asprintf
-      "type parameter `%s` has no solution. Maybe try an explicit instantiation."
-      (Cons.name c))) *)
+  error (Format.asprintf
+           "there is no consistent choice for type parameter `%s`."
```

Comment:
Is this message better then the one before?
This error is basically raised in situations like `expected 'Text' but got 'Nat'`. I'd leave the message in either form and improve in another PR
My idea was to start with this https://github.com/dfinity/motoko/pull/5640/files#diff-6d89983d022b561d8bcfe3f0add7c51eaca6df4d62dfd760cbb77468fc9ae187R7
However without incorporating the current solution-instantiation up until the error point the message looks wrong. That is why I'd like to solve the instantiation during an error and substitute. I continue working here https://github.com/dfinity/motoko/pull/5636